kebechet/managers/thoth_provenance/thoth_provenance.py

```diff
@@ -0,0 +1,107 @@
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+"""Consume Thoth Output for Kebechet auto-dependency management."""
+
+import hashlib
+import os
+import logging
+import json
+import typing
+import pprint
+
+from thamos import lib
+import git
+
+from kebechet.exception import DependencyManagementError
+from kebechet.exception import InternalError
+from kebechet.exception import PipenvError
+from kebechet.managers.manager import ManagerBase
+from kebechet.source_management import Issue
+from kebechet.source_management import MergeRequest
+from kebechet.utils import cloned_repo
+
+
+_LOGGER = logging.getLogger(__name__)
+
+_BRANCH_NAME = "kebechet_thoth"
+
+# Note: We cannot use pipenv as a library (at least not now - version 2018.05.18) - there is a need to call it
+# as a subprocess as pipenv keeps path to the virtual environment in the global context that is not
+# updated on subsequent calls.
+
+
+class ThothProvenanceManager(ManagerBase):
+    """Manage source issues of dependencies."""
+
+    def __init__(self, *args, **kwargs):
+        """Initialize ThothProvenance manager."""
+        # TODO: What needs to be changed for event driven Kebechet??
+        # We do API calls once for merge requests and we cache them for later use.
+        self._cached_merge_requests = None
+        super().__init__(*args, **kwargs)
+
+    @property
+    def sha(self):
+        """Get SHA of the current head commit."""
+        return self.repo.head.commit.hexsha
+
+    def _issue_provenance_error(self, prov_results: list, labels: list):
+        error_info = prov_results[0]
+        text_block = ""
+
+        for err in error_info:
+            LOGGER_.info(f"{err['id']: {err['package_name']} {err['package_version']}")
+            text_block = (
+                text_block
+                + f"## {err['type']}: {err['id']} - {err['package_name']}:{err['package_version']}\n"
+                + f"**Justification: {err['justification']}**\n"
+                + f"#### source: \n {pprint.pformat(err['source'])}\n"
+                + f"#### lock_info:\n {pprint.pformat(err['package_locked'])}"
+            )
+
+        checksum = hashlib.md5(text_block.encode("utf-8")).hexdigest()[:10]
+
+        self.sm.open_issue_if_not_exist(
+            f"{checksum} - {len(error_info)}: Automated kebechet thoth-provenance Issue",
+            lambda: text_block,
+            labels=labels,
+        )
+
+    def run(self, labels: list):
+        """Run the provenance check bot."""
+        with cloned_repo(self.service_url, self.slug, depth=1) as repo:
+            self.repo = repo
+            if os.path.isfile("Pipfile") and os.path.isfile("Pipfile.lock"):
+                res = lib.provenance_check_here()
+                for i in range(1, 11):
```

Comment:
I wonder - why is this needed?

Reply:
This is so that if the adviser failed, it will retry.  It has to force because otherwise it will repeatedly get the empty response.

Reply:
Kebechet should be simple here. If analysis failed, kebechet should report back it failed.